### ISSUE TITLE
Equal/NotEqual validators should perform ordinal string comparison by default

### ISSUE THREAD

**Issue by joshlang** [JeremySkinner/FluentValidation#1301]
Here's a lovely console app:

```
class Grr
{
    public string Asdf { get; set; } = "a\0";
}
class GrrValidator : AbstractValidator<Grr>
{
    public GrrValidator()
    {
        RuleFor(x => x.Asdf).Equal("a");
    }
}
class GrrValidator2 : AbstractValidator<Grr>
{
    public GrrValidator2()
    {
        RuleFor(x => x.Asdf).Equal("a", StringComparer.Ordinal);
    }
}

static void Main()
{
    Console.WriteLine(new GrrValidator().Validate(new Grr()).IsValid);
    Console.WriteLine(new GrrValidator2().Validate(new Grr()).IsValid);
}
```

The instance of `Grr` is constructed as two characters:  "a", then a \0 null terminator.

The validator checks to ensure the value equals "a".

The expected output is two lines saying `False`. However, we get
```
True
False
```

Unless we specify the string comparer, by default, one is chosen, probably according to the user's detected language/culture/religion/weight, in order to provide a nice localized experience.

But in our scenario, we don't want that.  

Can we set a default string comparer somehow?

I tried `ValidatorOptions.LanguageManager.Enabled = false;` without any luck.

We can always add StringComparer.Ordinal to every time we type out .Equal(..), but it's prone to us forgetting something, which leads to vulnerabilities for us.

**Comment by JeremySkinner** [comment]
> Unless we specify the string comparer, by default, one is chosen, probably according to the user's detected language/culture/religion/weight, in order to provide a nice localized experience.

That’s not intentional - it should be doing an ordinal comparison, but I suspect the Equal method is falling through to using IComparable.CompareTo, which does a culture-specific comparison. However, changing this might not be possible as this is likely to be a subtle breaking change for thousands of users - I’ll need to do some further research on the impact.

I can look at adding a way to provide a default comparer, but I’m probably not going to get a chance to work on this for some time. For now your best bet is to continue passing in the comparer as you’re doing. You could also wrap it in an extension method (EqualOrdinal or something like that) 

> I tried ValidatorOptions.LanguageManager.Enabled = false; without any luck.

This isn’t related to equality comparisons. The language manager is used for localising FluentValidation’s default error messages.

**Comment by joshlang** [comment]
Gotcha.  Okay, will do.  Thanks for a quick reply!

**Comment by JeremySkinner** [comment]
I did some further digging into this and it raised a whole load of other issues. 

I was correct that this is falling through to using string's `IComparable.CompareTo` method. This was introduced [by a contribution made to the project in 2012](https://github.com/JeremySkinner/FluentValidation/commit/fa1b6867d1fe6faf933ba73c27c69ce6d2e37f9c), which introduced the use of IComparable for Equal/NotEqual/ExclusiveBetween/InclusiveBetween/GreaterThan/GreaterThanOrEqual/LessThan/LessThanOrEqual

The problem is that `IComparable.CompareTo` is used for *sorting* not for equality comparison. I didn't catch this at the time I accepted the contribution, and so this issue has been in the codebase for over 7 years, affecting versions from 4.0 all the way to current (8.6.1).

The following test illustrates the difference between the two behaviours:

```csharp
[Fact]
public void Test() {
	string a = "a\0";
	string b = "a";

	Assert.False(a == b); // Ordinal
	Assert.False(a.Equals(b)); // Ordinal
       	Assert.True(a.Equals(b, StringComparison.CurrentCulture));
	Assert.False(object.Equals(a, b)); // Ordinal
	Assert.True(StringComparer.CurrentCulture.Equals(a, b));
	Assert.False(StringComparer.Ordinal.Equals(a, b));
	Assert.True(StringComparer.CurrentCulture.Compare(a, b) == 0);
	Assert.False(StringComparer.Ordinal.Compare(a, b) == 0);
	Assert.True(a.CompareTo(b) == 0); // Same as StringComparer.CurrentCulture.Compare
}

```
Internally the EqualValidator ends up using `a.CompareTo(b)`, and so runs through a culture-specific comparison intended for sorting, which is definitely not correct. 

So for strings specifically, I think this should change to a) not use IComparable b) do an ordinal comparison, but this is a subtle breaking change for anyone who may be relying on the existing behaviour, so I'm not sure if this would be better for a major release rather than a minor release. 

Ideally use of IComparable should be completely removed for all types as it's use is intended for sorting not for equality checks, but this is a bigger change and is less of an issue for non-string types.
